### PR TITLE
ddl: support modify column on partitioned table

### LINKED ISSUE #38297: Support MODIFY COLUMN of partitioned tables
## Feature Request

**Is your feature request related to a problem? Please describe:**

A partitioned table should support the same ALTER commands as a non-partitioned table. But things like [MODIFY COLUMN](https://docs.pingcap.com/tidb/v6.3/sql-statement-modify-column) or [CHANGE COLUMN](https://docs.pingcap.com/tidb/v6.3/sql-statement-change-column) does not support partitioned tables.

**Describe the feature you'd like:**

Support CHANGE COLUMN and MODIFY COLUMN on partitioned tables

**Describe alternatives you've considered:**

Create a new partitioned table with the changed column, copy all the data. But since there are no trigger support, it is really hard to  keep the new table in sync with the old one...

**Teachability, Documentation, Adoption, Migration Strategy:**

Same as any other CHANGE/MODIFY COLUMN use case, just for partitioned tables.

### PR DESCRIPTION
**Pull request by mjonss**
### What problem does this PR solve?

Issue Number: close #38297

Problem Summary:

Support for partitioned tables was never added for MODIFY COLUMN.

### What is changed and how it works?

Removed the limitation and added handling of partitions for the column change DDL operation. Updating indexes was already supported for partitioned tables, so only the table data / rows handling needed to be updated.

### Check List

Tests 

- [ ] Unit test
- [X] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [X] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
Support `ALTER TABLE t MODIFY column...` on partitioned table
```

### CODE REVIEW

File: ddl/failtest/fail_db_test.go [hunk 1]
```diff
@@ -494,7 +494,7 @@ func TestModifyColumn(t *testing.T) {
 
 	// Test unsupported statements.
 	tk.MustExec("create table t1(a int) partition by hash (a) partitions 2")
-	tk.MustGetErrMsg("alter table t1 modify column a mediumint", "[ddl:8200]Unsupported modify column: table is partition table")
+	tk.MustExec("alter table t1 modify column a mediumint")
```
Comment (Defined2014) [inline ddl/failtest/fail_db_test.go:497]: Maybe we could remove this? It's not an unsupported stmts any more.

**Review by Defined2014 — APPROVED** [review]

**Review by bb7133 — APPROVED** [review]
LGTM

**Pull request merged**